### ISSUE TITLE
2 bugs in `%` expansion

### ISSUE THREAD

**Issue by earthlyreason** [gittup/tup#478]
I love Tup.  I hate almost all software, but I love Tup.  If, like myself, the benevolent @gittup is now corrupted beyond the ability to tolerate any other build system (but, also like myself, still somehow tolerates building software), then I can hold out some hope that he may weigh in with some insight on this one.

Actually, there are two bugs here.  I am willing to help with these if there is anyone available to review my work.  (I am very inexperienced with C.)

For these examples, assume a directory with a zero-byte `Tupfile.ini` and a single `Tupfile` containing any of the example rules.

## Bug 1: `%` must be double-escaped in output section

The rule
```tup
: |> touch '%o' |> %%
```
incorrectly fails with
```
Unfinished %-flag at the end of the string '%'
```
It should create a file called `%`

### Observation (and workaround)

The rule
```tup
|> echo '%o' > '%o' |> %%%%
```
incorrectly “succeeds” by writing `%` to a file called `%`

It should have written `%%` to a file called `%%`

In other words, you can cope with this (well, for an even number of `%`'s) by quadrupling `%`'s (rather than doubling them) in the output section (modulo bug 2).

I have also confirmed that this affects extra outputs.  The rule

```tup
: |> touch '%o' '%%-2' |> dummy-1 | %%%%-2
```
incorrectly “succeeds” by creating `%-2`.

### Related cases that work correctly

In particular, `%%` works as expected in the process section (modulo bug 2).

- ✅  `|> echo '%%' > '%o' |> one-percent` correctly writes `%` to `one-percent`
- ✅  `|> touch '%o' |> %` correctly fails with `Unfinished %-flag at the end of the string '%'`
- ✅  `: |> echo '%' > '%o' |> one-percent` correctly fails with `%' must be followed by an 'f' for input files or an 'o' for output files.`
- ✅ `: |> echo '%o' > '%o' |> %%%` correctly fails with `Unfinished %-flag at the end of the string '%%%'`

I have not checked how `%%` is expanded in the input section.

## Bug 2: Certain forms involving `%` and `t` expose Tup internals

This bug is weirder.

In most cases, I find that you can “work around” bug 1 by double escaping `%` in the output section.

However, even this doesn't always work, due to another bug, in which the undocumented (and presumably unintended) form `%%#t` is interpreted as referencing certain Tup internals (where `#` is some numeral).

The easiest way to see what's going on here is with the following rules:

```tup
: |> echo '%%0t' > '%o' |> dummy-0t
: |> echo '%%1t' > '%o' |> dummy-1t
: |> echo '%%2t' > '%o' |> dummy-2t
: |> echo '%%3t' > '%o' |> dummy-3t
: |> echo '%%4t' > '%o' |> dummy-4t
: |> echo '%%5t' > '%o' |> dummy-5t
: |> echo '%%6t' > '%o' |> dummy-6t
: |> echo '%%7t' > '%o' |> dummy-7t
: |> echo '%%8t' > '%o' |> dummy-8t
: |> echo '%%9t' > '%o' |> dummy-9t
: |> echo '%%10t' > '%o' |> dummy-10t
: |> echo '%%11t' > '%o' |> dummy-11t
: |> echo '%%12t' > '%o' |> dummy-12t
: |> echo '%%13t' > '%o' |> dummy-13t
: |> echo '%%14t' > '%o' |> dummy-14t
: |> echo '%%15t' > '%o' |> dummy-15t
```
which, during processing will reveal their surprising expansions:

```
# 16) [0.008s] echo 'Tupfile.ini' > 'dummy-5t'                                                                             
# 15) [0.011s] echo '../{root}' > 'dummy-0t'                                                                               
# 14) [0.009s] echo 'Tupfile' > 'dummy-6t'                                                                                 
# 13) [0.008s] echo '^' > 'dummy-4t'                                                                                       
# 12) [0.008s] echo '.' > 'dummy-1t'                                                                                       
# 11) [0.009s] touch 'dummy-1' '%-2'                                                                                       
# 10) [0.009s] echo '/' > 'dummy-3t'                                                                                       
#  9) [0.009s] echo '$' > 'dummy-2t'                                                                                       
#  8) [0.002s] echo 'tup.config' > 'dummy-7t'                                                                              
#  7) [0.002s] echo '$/PATH' > 'dummy-8t'                                                                                  
#  6) [0.002s] echo '$/HOME' > 'dummy-9t'                                                                                  
#  5) [0.002s] echo 'one-percent' > 'dummy-10t'                                                                            
#  4) [0.002s] echo 'echo '%' > 'one-percent'' > 'dummy-11t'                                                               
#  3) [0.001s] echo '%' > 'dummy-12t'                                                                                      
#  2) [0.001s] echo 'dummy-1' > 'dummy-14t'                                                                                
#  1) [0.001s] echo 'echo '%' > '%'' > 'dummy-13t'                                                                         
#  0) [0.001s] echo '%-2' > 'dummy-15t'                                                                                    
```

All of these should expand the same way, as `%1t`, `%2t` and so on—as they do for every other letter of the alphabet besides lowercase `t`.

This bug also manifests in the output section when using the form `%%%%#t` pattern (which would be the workaround for bug 1), but usually as a crash with a misleading error message:

```
# 1 Not expecting '.' path here.
# 2 Attempting to insert '$' as a generated node when it already exists as a different type (directory).
# 3 Unable to write to a file outside of the tup hierarchy: /
# 4 %o used in rule pattern and no output files were specified.
# 5 Attempting to insert 'Tupfile.ini' as a generated node when it already exists as a different type (normal file).
# 6 Attempted to generate a file called 'Tupfile', which is reserved by tup.
# 7 Attempted to generate a file called 'tup.config', which is reserved by tup.
# 8 Attempting to insert '$/PATH' as a generated node when it already exists as a different type (@-variable).
# 9 Attempting to insert '$/HOME' as a generated node when it already exists as a different type (@-variable).
# 10 Unable to create output file 'one-percent' because it is already owned by command 11.
# 11 %' must be followed by an 'f' for input files or an 'o' for output files.
# 12 Unfinished %-flag at the end of the string '%'
# 13 %' must be followed by an 'f' for input files or an 'o' for output files.
# 14 Unable to create output file 'dummy-1' because it is already owned by command 16.
# 15 Unknown %-flag: '-'
# 16 Unknown %-flag: '-'
# 17 Unable to find node entry for tupid: 17
# 18 Unable to find node entry for tupid: 18
# (...presumably all the same...)
# 999 Unable to find node entry for tupid: 999
# : |> touch '%o' |> %%999t
```

**Comment by gittup** [comment]
I think the first bug is fixed in 772afcb102 - this came about because the strings have their %-flags expanded twice due to issue 257.

I'm not quite sure how to address your second issue. Yes, the %1t format is supposed to be an internal representation of an entry in the db for things like node variables (with the '&' syntax). I could probably put in an initial pass to make sure you don't type in %Nt somewhere in one of the strings, but it's possible you could still find ways to construct strings that eventually expand to that and get evaluated later. Are you noting this as a curiosity or is it actually causing problems for you?

**Comment by earthlyreason** [comment]
Thanks for the update.  I was running into this with a kind of "web cache" that was doing percent encoding of the filename to make URL's amenable to the file system.  Ultimately, I ended up using a different approach with a more controlled input domain, meaning it couldn't just take arbitrary web addresses, but you had to provide a key for each, which was acceptable.  I also changed to use Lua which I think avoids the first issue.

So, if the `%1t` behavior is a known thing, then I think there's nothing else at issue here.

Also p.s. sorry for the weird OP, I sound like I was on crack.
